Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool.swift

```diff
@@ -18,89 +18,102 @@ enum HTTPConnectionPool {
             case __testOnly_connection(ID, EventLoop)
         }
 
         private let _ref: Reference
 
-//        fileprivate static func http1_1(_ conn: HTTP1Connection) -> Self {
-//            Connection(_ref: .http1_1(conn))
-//        }
+        fileprivate static func http1_1(_ conn: HTTP1Connection) -> Self {
+            Connection(_ref: .http1_1(conn))
+        }
+
+        fileprivate static func http2(_ conn: HTTP2Connection) -> Self {
+            Connection(_ref: .http2(conn))
+        }
 
         static func __testOnly_connection(id: ID, eventLoop: EventLoop) -> Self {
             Connection(_ref: .__testOnly_connection(id, eventLoop))
         }
 
         var id: ID {
             switch self._ref {
-//            case .http1_1(let connection):
-//                return connection.id
+            case .http1_1(let connection):
+                return connection.id
+            case .http2(let connection):
+                return connection.id
             case .__testOnly_connection(let id, _):
                 return id
             }
         }
 
         var eventLoop: EventLoop {
             switch self._ref {
-//            case .http1_1(let connection):
-//                return connection.channel.eventLoop
+            case .http1_1(let connection):
+                return connection.channel.eventLoop
+            case .http2(let connection):
+                return connection.channel.eventLoop
             case .__testOnly_connection(_, let eventLoop):
                 return eventLoop
             }
         }
 
-        @discardableResult
-        fileprivate func close() -> EventLoopFuture<Void> {
+        fileprivate func execute(request: HTTPExecutableRequest) {
             switch self._ref {
-//            case .http1_1(let connection):
-//                return connection.close()
-
-            case .__testOnly_connection(_, let eventLoop):
-                return eventLoop.makeSucceededFuture(())
+            case .http1_1(let connection):
+                return connection.execute(request: request)
+            case .http2(let connection):
+                return connection.executeRequest(request)
+            case .__testOnly_connection:
+                break
             }
         }
 
-        fileprivate func execute(request: HTTPExecutableRequest) {
+        /// Shutdown cancels any running requests on the connection and then closes the connection
+        fileprivate func shutdown() {
             switch self._ref {
-//            case .http1_1(let connection):
-//                return connection.execute(request: request)
+            case .http1_1(let connection):
+                return connection.shutdown()
+            case .http2(let connection):
+                return connection.shutdown()
             case .__testOnly_connection:
                 break
             }
         }
 
-        fileprivate func cancel() {
+        /// Closes the connection without cancelling running requests.
```

Comment:
What happens to the requests if we don't cancel them? Why would do use this over shutdown?

Reply:
In the pool state machine, we know if a connection is idle or not. For this reason, we can use the direct way to close, if we are sure nothing is running on the connection.